### ISSUE TITLE
Passing a test dataset to KernelSHAP

### ISSUE THREAD

**Issue by varunbanda** [SeldonIO/alibi#339]
Hi,

I have a question regarding the explainer (and subsequent shap values) that is created using KernelSHAP.

We always pass the training/subset/kmeans cluster for KernelSHAP.

I would like to know what are the implications of passing a test dataset (unseen data for the trained model) instead of the train dataset. I have experimented using the wine dataset and found out that the feature importances are almost the same (with 1 or 2 features trading places - mean absolute SHAP values were pretty close).

Is this because the test dataset that I had passed is very similar to the train dataset?

**Comment by jklaise** [comment]
Yes, as long as the distributions between train/test are the similar, then you can expect the SHAP values to be the same (within random error).

The distinction between the train/test set for using SHAP is similar to any ML task. For a modelling task you would eventually select "the best" model (based on cross-validation or similar) and then use it to retrain on all of your available data before deploying, so in effect all of your data becomes the training set at this point. Same with SHAP, you should use the same set of data as for your final trained model which usually means all of the data that's available.

If you start including more data in fitting SHAP than what your model has seen then it becomes trickier to interpret any SHAP values derived from this procedure as this new data could no longer be representative of the data the model has seen.

**Comment by varunbanda** [comment]
Thanks for the detailed explanation. Really appreciate it!
